src/MetricsUploader/CaptureMetricTest.cpp

```diff
@@ -0,0 +1,159 @@
+              (override));
+  MOCK_METHOD(bool, SendLogEvent,
+              (OrbitLogEvent_LogEventType /*log_event_type*/,
+               std::chrono::milliseconds /*event_duration*/,
+               OrbitLogEvent_StatusCode /*status_code*/),
+              (override));
+  MOCK_METHOD(bool, SendCaptureEvent,
+              (OrbitCaptureData /*capture data*/, OrbitLogEvent_StatusCode /*status_code*/),
+              (override));
+};
+
+TEST(CaptureMetric, SendEmpty) {
+  MockUploader uploader{};
+  CaptureMetric metric{&uploader, kTestStartData};
+  EXPECT_FALSE(metric.Send());
+}
+
+TEST(CaptureMetric, SetCaptureFailedAndSend) {
+  MockUploader uploader{};
+
+  EXPECT_CALL(uploader, SendCaptureEvent(_, _))
+      .Times(1)
+      .WillOnce(
+          [](const OrbitCaptureData& capture_data, OrbitLogEvent_StatusCode status_code) -> bool {
+            EXPECT_EQ(status_code, OrbitLogEvent_StatusCode_INTERNAL_ERROR);
+            EXPECT_GE(capture_data.duration_in_milliseconds(), 5);
+            EXPECT_TRUE(CaptureDataProtoContainsCaptureStartData(capture_data, kTestStartData));
+            return true;
+          });
+
+  CaptureMetric metric{&uploader, kTestStartData};
+  std::this_thread::sleep_for(std::chrono::milliseconds(5));
+  metric.SetCaptureFailed();
+  EXPECT_TRUE(metric.Send());
+}
+
+TEST(CaptureMetric, SetCaptureCancelledAndSend) {
+  MockUploader uploader{};
+
+  EXPECT_CALL(uploader, SendCaptureEvent(_, _))
+      .Times(1)
+      .WillOnce(
+          [](const OrbitCaptureData& capture_data, OrbitLogEvent_StatusCode status_code) -> bool {
+            EXPECT_EQ(status_code, OrbitLogEvent_StatusCode_CANCELLED);
+            EXPECT_GE(capture_data.duration_in_milliseconds(), 5);
+            EXPECT_TRUE(CaptureDataProtoContainsCaptureStartData(capture_data, kTestStartData));
+            return true;
+          });
+
+  CaptureMetric metric{&uploader, kTestStartData};
+  std::this_thread::sleep_for(std::chrono::milliseconds(5));
+  metric.SetCaptureCancelled();
+  EXPECT_TRUE(metric.Send());
+}
+
+TEST(CaptureMetric, SetCaptureCompleteAndSend) {
+  MockUploader uploader{};
+
+  CaptureCompleteData complete_data{
+      std::chrono::milliseconds{51} /*duration_in_milliseconds*/
+  };
+
+  EXPECT_CALL(uploader, SendCaptureEvent(_, _))
+      .Times(1)
+      .WillOnce([complete_data](const OrbitCaptureData& capture_data,
+                                OrbitLogEvent_StatusCode status_code) -> bool {
+        EXPECT_EQ(status_code, OrbitLogEvent_StatusCode_SUCCESS);
+        EXPECT_EQ(capture_data.duration_in_milliseconds(),
+                  complete_data.duration_in_milliseconds.count());
+        EXPECT_TRUE(CaptureDataProtoContainsCaptureStartData(capture_data, kTestStartData));
+        return true;
+      });
+
+  CaptureMetric metric{&uploader, kTestStartData};
+  metric.SetCaptureComplete(complete_data);
+  EXPECT_TRUE(metric.Send());
+}
+
+TEST(CaptureMetric, MultipleSetAndSend) {
```

Comment:
This could be a real situation? Why we are testing this?

Reply:
No, currently this is not used. However, the class `CaptureMetric` might be used by someone else in the future (maybe ClientGgp). So I designed the class in a way that the outcome of these setters is always the same. The setters basically set the state of the class. Is is confusing for users of the class, if a call to a setter has a different outcome depending on the state of the class before the call. So I wrote this test to make sure the class adheres to this. I hope this explanation makes a little bit of sense. 

The idea actually came from @beckerhe (feel free to chime in).

Reply:
mmm, in this case I will rename the test to CheckingSameOutcomeSetAndSend and maybe 3 similar CaptureMetrics metric1, metric2 and metric3?

Reply:
I improved the test